src/applications/applications.service.ts

```diff
@@ -249,90 +249,115 @@ export class ApplicationsService {
-            row.push((i + 1).toString());
-          } else {
-            row.push('');
-          }
-        }
-
-        // Application data fields
-        for (const field of appDataFields) {
-          if (field === 'otr') {
-            row.push(app.applicationData['nspOtr'] !== undefined ? app.applicationData['nspOtr'] : '');
-          } else if (field === 'aadhaar') {
-            const aadhaar = app.applicationData['aadhaar'];
-            row.push(aadhaar ? aadhaar.slice(-4) : '');
-          } else {
-            row.push(app.applicationData[field] !== undefined ? app.applicationData[field] : '');
-          }
-        }
-
-        // Application table data fields
-        for (const field of applicationTableDataFields) {
-          if (field === 'amount') {
-            row.push(app.finalAmount ?? '');
-          } else if (field === 'applicationId') {
-            row.push(app.id !== undefined ? app.id : '');
-          } else {
-            row.push(app[field] !== undefined ? app[field] : '');
-          }
+  private resolveDynamicFields(apps: any[], fields: string[], source: 'applicationData' | 'calculatedAmount'): string[] {
+    if (fields.length === 1 && fields[0] === '*') {
+      const keySet = new Set<string>();
+      for (const app of apps) {
+        const sourceData = app[source];
+        if (sourceData && typeof sourceData === 'object') {
+          Object.keys(sourceData).forEach(key => keySet.add(key));
         }
-
-        csvRows.push(row.map(val => `"${String(val).replace(/"/g, '""')}"`).join(','));
       }
+      return Array.from(keySet);
+    }
+    return fields;
+  }
 
-      return csvRows.join('\n');
-    };
+  private generateAutoFields(fields: string[], index: number): (string | number)[] {
+    return fields.map(field => field === 'serialNumber' ? index + 1 : '');
+  }
 
-    let dynamicAppDataFields: string[] = [];
-    if (applicationDataColumnDataFields.length === 1 && applicationDataColumnDataFields[0] === '*') {
-      // Fetch applications first to get all keys
-      const fieldSet = new Set<string>();
-      for (const app of applications) {
-        if (app.applicationData && typeof app.applicationData === 'object') {
-          Object.keys(app.applicationData).forEach(key => fieldSet.add(key));
-        }
-      }
-      dynamicAppDataFields = Array.from(fieldSet);
+  private generateAppDataFields(app: any, fields: string[]): string[] {
+    return fields.map(field => {
+      if (field === 'otr') return app.applicationData?.nspOtr ?? '';
+      if (field === 'aadhaar') return app.applicationData?.aadhaar?.slice(-4) ?? '';
+      return app.applicationData?.[field] ?? '';
+    });
+  }
 
-      const headerFields = [...autoGenerateFields, ...dynamicAppDataFields, ...applicationTableDataFields];
-      return generateCsvRows(applications, headerFields, dynamicAppDataFields);
-    }
+  private generateCalcAmountFields(app: any, fields: string[]): string[] {
```

Comment:
These field values and CSV labels should get generated runtime based on calculation rules as calculation rules may change scheme to scheme and could have any name @milindSansare